### PR TITLE
Added docs for fp16

### PR DESCRIPTION
**Pull request by marii-moe**
Docs for the fp16 related callbacks

### CODE REVIEW

**Comment by marii-moe** [comment]
This is for #45

File: fastai/callback/fp16.py [hunk 1]
```diff
@@ -114,14 +134,22 @@ def after_fit (self): self.learn.model = convert_network(self.model, dtype=torch
...
-                 div_factor=2., scale_wait=500, clip=None):
+    def __init__(self,
+        loss_scale:int=512, # Loss scale to use if not using dynamic
+        flat_master:bool=False, # Whether to flatten fp32 parameters for performance
+        dynamic:bool=True, # Whether to automatically determine loss scaling
+        max_loss_scale:float=2.**24, # Starting value for dynamic loss scaling
+        div_factor:float=2., # Divide by this on overflow, multiply by this after scale_wait batches
+        scale_wait:int=500, # Number of batches to wait for increasing loss scale
+        clip:float=None, # Value to clip gradients at, max_norm, as in `nn.utils.clip_grad_norm_`
+):
```
Comment (warner-benjamin) [inline fastai/callback/fp16.py:145]:
```suggestion
    ):
```

File: fastai/callback/fp16.py [hunk 2]
```diff
@@ -57,7 +62,10 @@ def to_fp32(self:Learner): return self.remove_cb(MixedPrecision)
 from torch.nn.utils import parameters_to_vector
 
 # Cell
-def get_master(opt, flat_master=False):
+def get_master(opt:Optimizer, # Optimizer from which to retrieve model params
+    flat_master:bool=False, # Flatten fp32 params into a vector for better performance
+) -> [[[nn.Parameter]],[[nn.Parameter]]]: # List of fp16 params, and list of fp32 params
```
Comment (warner-benjamin) [inline fastai/callback/fp16.py:67]:
```suggestion
def get_master(
    opt:Optimizer, # Optimizer from which to retrieve model params
    flat_master:bool=False, # Flatten fp32 params into a vector for better performance
) -> [[[nn.Parameter]],[[nn.Parameter]]]: # List of fp16 params, and list of fp32 params
```

File: fastai/callback/fp16.py [hunk 3]
```diff
@@ -71,22 +79,34 @@ def get_master(opt, flat_master=False):
...
+):
+    "Tests whether fp16 gradients have overflowed."
     s = float(x.float().sum())
     return (s == float('inf') or s == float('-inf') or s != s)
 
 # Cell
-def grad_overflow(pgs):
+def grad_overflow(pgs:[[nn.Parameter]] # Fp16 parameter groups to test for overflow
+)->bool:
+    "Tests all fp16 parameters in pgs for gradient overflow"
```
Comment (warner-benjamin) [inline fastai/callback/fp16.py:109]: I am under the impression that for these one argument methods we can expand the doc string if needed in leu of a docment
```suggestion
def grad_overflow(pgs:[[nn.Parameter]])->bool:
    "Tests all fp16 parameters in pgs for gradient overflow"
```

**Review by warner-benjamin — CHANGES_REQUESTED** [review]
Some minor style changes needed. I provided an example using `get_master`.

Assuming I am correct in my assumption, `grad_overflow` could be formatted as my suggestion, if you like.

@muellerzr wanted to take a look at this one, so added him as a reviewer.

All the docment comments look good to me.

File: fastai/callback/fp16.py (hunk 1 above)
Comment (marii-moe) [reply fastai/callback/fp16.py:145]: ugh, I probably spent a good 15 minutes staring at this wondering why jupyter was formatting it the way you want here, thinking jupyter was broken, mystery solved :)

**Review by warner-benjamin — APPROVED** [review]
Looks good.

File: nbs/18_callback.fp16.ipynb [hunk 4]
```diff
@@ -202,18 +202,49 @@
     "    def after_loss(self): self.autocast.__exit__(None, None, None)\n",
```
Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #15.            "Use self as a fake optimizer. self.skipped will be set to True after_step if gradients overflow. "
</pre><p>Q: Is the optimizer step called after this step? or this is the actual optimizer step?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #24.            "Pretend to be an optimizer for <code>GradScaler</code>"
</pre><p>Not so sure, but I wonder, if this class delegates things to <code>GradScaler</code>  isnt more like just "Pretend to be an optmizer which has param_groups"... so it is not being like an optimizer for gradScaler, but to whoeveer is calling this property on a <code>MixedPrecision</code>  Instance?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='17'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #4.        flat_master:bool=False, # Flatten fp32 params into a vector for better performance
</pre><p>Better performance and half memory? or that last is not always true?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #4.        master_pgs:[[nn.Parameter]], # Fp32 model parameters to copt gradients to
</pre><p>typo in <code>copt</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #3.        "Tests whether fp16 gradients have overflowed."
</pre><p>Could be <code>Tests whether fp16 gradients have overflow.</code>?  or maybe <code>are overflowed</code>  (really dont know as I fail on english here).</p><p> </p><p>I Also see some docstrings with an ending <code>.</code>  and others without it, it seems that all docstrings should end with a dot.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #7.            loss_scale:int=512, # Loss scale to use if not using dynamic
</pre><p>Is there an easy explanation of what loss scale is or why it is 512 by default?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (tyoc213) [inline nbs/18_callback.fp16.ipynb:202]: <pre class="ql-syntax" spellcheck="false">Line #11.            div_factor:float=2., # Divide by this on overflow, multiply by this after scale_wait batches
</pre><p><code>test_overflow</code> is the one that directs this "on overflow"? (just curious)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/muellerzr/fastai-docment-sprint/pull/70/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='56'/>

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: No, it is a lot more nuanced. This would flatten the fp32 wights into a single vector so we can update them all at once as opposed to in a loop.

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: Ah, "overflown" is probably the correct word.

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: Good point, I don't really attempt to explain loss scale here. 512 is a fairly good default, I usually see dynamic settling at 256-512 for good values. Generally people shouldn't use this and instead use dynamic loss scale.

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: Yes if test_overflow finds the gradients have overflown, then we will divide the loss scale by this number. Otherwise we multiply by this number is we do not get an overflow after scale_wait number of batches.

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: The optimizer step is called after before_step. In before_step we fake out the GradScaler in this like: `self.scaler.step(self)`. The GradScaler expects an optimizer here. If we overflow here, then the batch is canceled and the actual Optimizer.step is never called for this batch.

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: The Learner still uses its normal optimizer. We fake out the GradScaler in before_step, specifically this line: `self.scaler.step(self)` where GradScaler expects us to pass in an optimizer.

Comment (tyoc213) [reply nbs/18_callback.fp16.ipynb:202]: OK, so no problem

Comment (tyoc213) [reply nbs/18_callback.fp16.ipynb:202]: so it should be a power of 2? or it can be any number like 17?

Comment (marii-moe) [reply nbs/18_callback.fp16.ipynb:202]: Could be anything, but is generally a multiple of 2, because div_factor is by default set to be 2. Truthfully, I don't know anyone who would actually have a legitimate use case for actually using most of these settings, and I would expect them to instead simply use default dynamic scaling.

**Review by warner-benjamin — COMMENTED** [review]
Looks good but a couple of caveats. 

The `NonNativeMixedPrecision._docs` dict might overwrite the new doc strings which were added. Some of the preexisting doc strings via `_docs` are more detailed than the added doc strings.

**Comment by marii-moe** [comment]
@warner-benjamin updated given your suggestion

@tyoc213 could you take another look?

**Comment by tyoc213** [comment]
@marii-moe lg2m

**Review by tyoc213 — APPROVED** [review]

**Comment by marii-moe** [comment]
@muellerzr you said you wanted to review this one about 2 weeks ago.
